op-chain-ops/foundry/sourcefs.go

```diff
@@ -0,0 +1,147 @@
+package foundry
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/fs"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"golang.org/x/exp/maps"
+
+	"github.com/ethereum-optimism/optimism/op-chain-ops/srcmap"
+)
+
+// SourceMapFS wraps an FS to provide source-maps.
+// This FS relies on the following file path assumptions:
+// - `/artifacts/build-info/X.json` (build-info path is read from the below file): build files, of foundry incremental builds.
+// - `/cache/solidity-files-cache.json`: a JSON file enumerating all files, and when the build last changed.
+// - `/` a root dir, relative to where the source files are located (as per the compilationTarget metadata in an artifact).
+type SourceMapFS struct {
+	fs fs.FS
+}
```

Comment:
I had assumed you'd be using the `//go:embed` directive to handle embedded sources. This would allow you to get all the source files for a given directory, and they'd even be available in the compiled binary if that's applicable here.

There's a lot to consider with what you've put together so I'm not sure if an embedded FS would have been a big help, but was this considered and decided against?

Reply:
An embedded FS is compatible with this, it can be plugged in. Packaging the contracts data is a different problem from consuming the package data. And while packaging the artifacts might be useful later on, packaging the source-code is unlikely, since it's a lot of data that we don't want to tightly couple. But this way with the FS it's very optional and flexible in the way we plug it in.